domain/model/state/state.go

```diff
@@ -793,71 +788,178 @@ func (s *State) ListAllModels(ctx context.Context) ([]coremodel.Model, error) {
+	excludeIDs []coremodel.UUID,
+) ([]coremodel.HostedModel, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	query := `
+SELECT &dbModel.*
+FROM v_model
+WHERE life IN ($lifeList[:])
+AND uuid NOT IN ($modelIDList[:])
+`
+
+	inputLifes := lifeList(includeLifes)
+	inputModelIDs := modelIDList(excludeIDs)
+
+	stmt, err := sqlair.Prepare(query, dbModel{}, inputLifes, inputModelIDs)
+	if err != nil {
+		return nil, fmt.Errorf("preparing get model statement: %w", err)
+	}
+
+	resultsForModel := []dbModel{}
+	modelClouds := map[string]cloud.Cloud{}
+	modelCredentials := map[string]cloud.Credential{}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, inputLifes, inputModelIDs).GetAll(&resultsForModel)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return fmt.Errorf("getting models from database: %w", err)
+		}
+
+		for _, stateModel := range resultsForModel {
+			modelCloud, err := cloudstate.GetCloudForID(ctx, s, tx, corecloud.ID(stateModel.CloudID))
 			if err != nil {
-				return err
+				return fmt.Errorf("cannot get cloud %q for model %q: %w", stateModel.CloudID, stateModel.UUID, err)
 			}
+			modelClouds[stateModel.UUID] = modelCloud
 
-			model.CloudRegion = cloudRegion.String
-			model.ModelType = coremodel.ModelType(modelType)
-			model.Owner = user.UUID(userUUID)
-			model.Credential = credKey
-			rval = append(rval, model)
+			modelCredential, err := credentialstate.GetCloudCredential(ctx, s, tx, corecredential.ID(stateModel.CredentialID))
+			if err != nil {
+				return fmt.Errorf("cannot get credential %q for model %q: %w", stateModel.CredentialID, stateModel.UUID, err)
+			}
+			modelCredentials[stateModel.UUID] = modelCredential.AsCredential()
 		}
 
-		return rows.Err()
+		return nil
 	})
+	if err != nil {
+		return nil, fmt.Errorf("getting all models: %w", err)
+	}
+
+	hostedModels := make([]coremodel.HostedModel, 0, len(resultsForModel))
+	for _, stateModel := range resultsForModel {
+		coreModel, err := stateModel.toCoreModel()
+		if err != nil {
+			return nil, fmt.Errorf("transforming model %q from db: %w", stateModel.UUID, err)
+		}
+
+		hostedModels = append(hostedModels, coremodel.HostedModel{
+			Model:      coreModel,
+			Cloud:      modelClouds[stateModel.UUID],
+			Credential: modelCredentials[stateModel.UUID],
+		})
+	}
+	return hostedModels, nil
+}
+
+// ModelLastLogins lists all models along with the last login by the specified
+// user. This function will filter on models that have a "life" value in the
+// includeLifes slice. If no matching models exist, an empty slice is returned.
+func (s *State) ModelLastLogins(
+	ctx context.Context,
+	userID user.UUID,
+	includeLifes []corelife.Value,
+) ([]coremodel.ModelWithLogin, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	query := `
+SELECT &dbModel.*, &userModelLastLogin.*
+FROM v_model
+LEFT JOIN model_last_login AS mll
```

Comment:
Watch out for empty last logins here with the left join. If the user has never logged in to the model the last login columns will be empty. You have a `*time.Time` in the `userModelLastLogin` struct so that will just be `nil`, but the `UserID` in that will be empty in this case and I don't think you have a check for that.

Reply:
yeah the `UserID` is not actually being used, we could remove it